src/util/json/PrototypeExpander.js

```diff
@@ -0,0 +1,1229 @@
+				}
+				// console.error("Done searching, found", nuobject);
+				done(p, nuobject, protoexp);
+			});
+		} catch (e) {
+			console.error("Error searching for proto", e);
+		}
+	},
+
+	/**
+	 * processURLs and make them more kosher for the X3DJSONLD user inteferface to
+	 * deal with.  Pass an array of URLs and a path for the main JSON file you are
+	 * loading.
+	 */
+	processURLs: function(localArray, path) {
+		// console.error("Process URLs", path, localArray);
+		var url;
+		// No longer need to split
+		for (url in localArray) {
+			if (localArray[url].indexOf("http://") === 0
+			 || localArray[url].indexOf("https://") === 0) {
+			} else if (localArray[url].indexOf("urn:web3d:media:textures/panoramas/") === 0) {
+				var ls = localArray[url].lastIndexOf("/");
+				if (ls > 0) {
+					localArray[url] = 'examples/Basic/UniversalMediaPanoramas/'+localArray[url].substring(ls+1);
+				}
+
+			} else {
+				/*
+				var s = localArray[url].indexOf('/');
+				*/
+				var p = localArray[url].indexOf('#');
+				var pe = path.lastIndexOf('/');
+				var pc = path;
+				if (pe >= 0) {
+					pc = path.substring(0, pe);
+				}
+				/*
+				if (s != 0 && p != 0) {
+					if (localArray[url].indexOf(pc) != 0) {
+						 localArray[url] = pc+'/'+localArray[url];
+					}
+					if (localArray[url].indexOf('/') === 0) {
+						// no webroot absolute paths.  No /'s for X_ITE shaders
+						localArray[url] = localArray[url].substring(1);
+					}
+				}
+				*/
+				while (localArray[url].startsWith("../")) {
+					localArray[url] = localArray[url].substr(3);
+					var pe = pc.lastIndexOf('/');
+					if (pe >= 0) {
+						pc = pc.substring(0, pe);
+					} else {
+						pc = "";
+					}
+				}
+				if (p == 0) {
+					localArray[url] = path+localArray[url];
+				} else {
+					localArray[url] = pc+"/"+localArray[url];
+				}
+			}
+			// for server side
+			var h = localArray[url].lastIndexOf("#") ;
+			var hash = "";
+			if (h >= 0) {
+				hash = localArray[url].substring(h);
+				localArray[url] = localArray[url].substring(0, h);
+			}
+			/*
+			var x3d = localArray[url].lastIndexOf(".x3d") ;
+			if (x3d === localArray[url].length - 4) {
+				localArray[url] = localArray[url].substring(0, x3d)+".json" + hash;
+			}
+			*/
+			var wrl = localArray[url].lastIndexOf(".wrl") ;
+			if (wrl === localArray[url].length - 4) {
+				localArray[url] = localArray[url].substring(0, wrl)+".json" + hash;
+			}
+			var wrz = localArray[url].lastIndexOf(".wrz") ;
+			if (wrz === localArray[url].length - 4) {
+				localArray[url] = localArray[url].substring(0, wrz)+".json" + hash;
+			}
+				
+		}
+		// console.error("Processed URLs", localArray.join(" "));
+		return localArray;
+	},
+
+	/**
+	 * Use almost any method possible to load a set of URLs.  The loadpath is the
+	 * original URL the main JSON got laoded from, Urls is the se of urls, and
+	 * the loadedCallback returns the data and the URL it was loaded from.
```

Comment:
Perhaps add that the http(s) and fs modules are required on server with node.js ?